singularity-eos/eos/eos_mgusup.hpp

```diff
@@ -0,0 +1,365 @@
+  Real f1 = 1.0 - _s * eta;
+  if (f1 <= 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("MGUsup model parameters s and rho0 together with rho "
+                                   "give a negative argument for a logarithm.");
+  }
+  Real G0os = robust::ratio(_G0, _s);
+  // sk, lk, mk, sum, and enough values may change in the loop
+  Real sk = -_G0 * eta;
+  Real lk = _G0;
+  Real mk = 0.0;
+  Real sum = sk;
+  int enough = 0;
+  Real temp;
+  Real pf = _Cs * _Cs / (2.0 * _Cv0 * _s * _s);
+  if (eta * eta < 1.e-8) {
+    temp = _T0 * exp(_G0 * eta) +
+           pf * (2.0 * std::log(f1) - _s * eta / (f1 * f1) * (3.0 * _s * eta - 2.0));
+  } else {
+    for (int i = 0; ((i < sumkmax) && (enough == 0)); i++) {
+      mk = G0os / (i + 2) / (i + 2);
+      sk = (sk * f1 - lk * eta) * (i + 1) * mk;
+      lk = mk * (i + 1) * lk;
+      sum = sum + sk;
+      if (robust::ratio((sk * sk), (sum * sum)) < (cutoff * cutoff)) {
+        enough = 1;
+      }
+    }
+    if (enough == 0) {
+#ifndef NDEBUG
+      PORTABLE_WARN("Hugoniot Temperature not converged");
+#endif // NDEBUG
+    }
+    // printf("sum=%e\n",sum);
+    temp = _T0 - pf * ((G0os - 3.0) + exp(-G0os) * (G0os * G0os - 4.0 * G0os + 2.0) *
+                                          (std::log(f1) + sum));
+    temp = temp * exp(_G0 * eta);
+    temp = temp - pf / f1 * ((4.0 - G0os) - 1.0 / f1);
+  }
+  return temp;
+}
+
+PORTABLE_INLINE_FUNCTION Real MGUsup::InternalEnergyFromDensityTemperature(
+    const Real rho, const Real temp, Real *lambda) const {
+  Real value =
+      HugInternalEnergyFromDensity(rho) + _Cv0 * (temp - HugTemperatureFromDensity(rho));
+  return value;
+}
+PORTABLE_INLINE_FUNCTION Real MGUsup::PressureFromDensityTemperature(const Real rho,
+                                                                     const Real temp,
+                                                                     Real *lambda) const {
+  Real value = HugPressureFromDensity(rho) +
+               _G0 * _rho0 * _Cv0 * (temp - HugTemperatureFromDensity(rho));
+  return value;
+}
+PORTABLE_INLINE_FUNCTION Real MGUsup::EntropyFromDensityTemperature(const Real rho,
+                                                                    const Real temp,
+                                                                    Real *lambda) const {
+  Real eta = 1.0 - robust::ratio(_rho0, rho);
+  Real value = _S0 - _G0 * _Cv0 * eta + _Cv0 * std::log(temp / _T0);
+  return value;
+}
+PORTABLE_INLINE_FUNCTION Real MGUsup::TExpansionCoeffFromDensityTemperature(
+    const Real rho, const Real temp, Real *lambda) const {
+  Real value =
+      robust::ratio(_Cv0 * _rho0 * _G0, TBulkModulusFromDensityTemperature(rho, temp));
+  return value;
+}
+PORTABLE_INLINE_FUNCTION Real MGUsup::TBulkModulusFromDensityTemperature(
+    const Real rho, const Real temp, Real *lambda) const {
+  Real eta = 1.0 - robust::ratio(_rho0, rho);
+  Real value = robust::ratio((1.0 + _s * eta - _G0 * _s * eta * eta), (1.0 - _s * eta));
```

Comment:
Similarly to the above comment, I think this would lead to a negative bulk modulus when the maximum compression limit is exceeded.

Reply:
This will throw the code since it uses HugTemperatureFromDensity where this would be caught. We leave the discussion on how to do this properly to later after resolving the issue via Mattermost and/or other channels.